app/scripts/controllers/transactions/tx-state-manager.test.js

```diff
@@ -614,6 +631,136 @@ describe('TransactionStateManager', function () {
+      assert.notEqual(result[0].id, 0, 'first tx should be removed');
+      assert.notEqual(
+        result[0].status,
+        TRANSACTION_STATUSES.DROPPED,
+        'first tx should not be dropped',
+      );
+      assert.notEqual(
+        result[1].type,
+        TRANSACTION_TYPES.CANCEL,
+        'Cancel tx should be dropped',
+      );
+    });
+
+    it('cuts off entire groups of transactions by nonce + network when adding new transaction', function () {
+      const limit = txStateManager.txHistoryLimit;
+      // In this test case the earliest two transactions are a dropped attempted ether send and a
+      // following cancel transaction with the same nonce. Then, a bit later the same scenario on a
+      // different network. The first two transactions should be dropped after adding even another
+      // single transaction but the other shouldn't be dropped until adding the fifth additional
+      // transaction
+      const txs = generateTransactions(limit + 5, {
+        chainId: (i) => ([0, 1, 4, 5].includes(i) ? currentChainId : '0x1'),
+        to: VALID_ADDRESS,
+        from: VALID_ADDRESS_TWO,
+        nonce: (i) => {
+          if (i === 1) return '0';
+          else if (i === 5) return '4';
+          return `${i}`;
+        },
+        status: (i) =>
+          i === 0 || i === 4
+            ? TRANSACTION_STATUSES.DROPPED
+            : TRANSACTION_STATUSES.CONFIRMED,
+        type: (i) =>
+          i === 1 || i === 5
+            ? TRANSACTION_TYPES.CANCEL
+            : TRANSACTION_STATUSES.SENT_ETHER,
+      });
+      txs.forEach((tx) => txStateManager.addTransaction(tx));
+      const result = txStateManager.getTransactions({
+        filterToCurrentNetwork: false,
+      });
+      assert.equal(
+        result.length,
+        limit + 1,
+        `limit of ${limit} + 1 for the grouped transactions is enforced`,
+      );
+      assert.notEqual(result[0].id, 0, 'first tx should be removed');
+      assert.equal(
+        result[0].status,
+        TRANSACTION_STATUSES.DROPPED,
+        'first tx should be dropped',
+      );
+      assert.equal(
+        result[0].txParams.nonce,
+        '0x4',
+        'the first tx should be from the second group',
+      );
+      assert.equal(
+        result[1].type,
+        TRANSACTION_TYPES.CANCEL,
+        'second transaction should be a cancel',
+      );
+      assert.equal(
+        result[1].txParams.nonce,
+        '0x4',
+        'the second tx should be from the second group',
+      );
+    });
+
+    it('does not cut off entire groups of transactions when adding new transaction when under limit', function () {
+      // In this test case the earliest two transactions are a dropped attempted ether send and a
+      // following cancel transaction with the same nonce. Then, a bit later the same scenario on a
+      // different network. None of these should be dropped because we haven't yet reached the limit
+      const txs = generateTransactions(9, {
```

Comment:
Nit: it would probably be better to express this in terms of the limit. (e.g. `limit - 1`), so that it's more clear why this number was chosen, and so that it continues to work if the limit changes.

Reply:
done in latest